# EDA Project: on Google Play Store Data with 2.3+ Millions Records 

Internship Platform: Data Glacier, [KultureHire](https://www.kulturehire.com/candidate/dashboard) | [internstreet](https://www.internstreet.co/)
https://www.kulturehire.com/candidate/dashboard

## Complete Exploratory Data Analysis
---
### About Dataset
> **Description**: This dataset in taken from kaggle from following [link](https://www.kaggle.com/datasets/gauthamp10/google-playstore-apps?select=Google-Playstore.csv)

Gautham Prakash, "Google Play Store App data of 2.3 Million+ applications." June 2021 [Online] [Available](https://www.kaggle.com/datasets/gauthamp10/google-playstore-apps?select=Google-Playstore.csv)

**Context**:

Google PlayStore Android App Data. (2.3 Million+ App Data)\
Backup repo: [Link](https://github.com/gauthamp10/Google-Playstore-Dataset)

**Content**:

I've collected the data with the help of Python script (Scrapy) running on a cloud vm instance.
The data was collected in the month of June 2021.

**Acknowledgements**:

I couldn't have build this dataset without the help of Github Education and switched to facundoolano/google-play-scraper for sane reasons.

**Inspiration**:

Took inspiration from: [Link](https://www.kaggle.com/lava18/google-play-store-apps) to build a big database for students and researchers.


## Step 1: Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
%matplotlib inline

## Step 2:  Data Loading and Exploration | Cleaning
- Loading the Dataset
- Storing the data into Dataframe to apply pandas library
- Dealing with missing data, outliers, inconsistancies in dataset

- let's load the dataset

In [2]:
# loading dataset
df2 = pd.read_csv('./data/Google-Playstore.csv')
# df2 = df.sample(frac = 0.25)

- I need to sample the data first for fast processing

In [3]:
df = df2.sample(frac = 0.1)

> Note: Dataset's all columns and rows does not presents for observations so we some important commands to show the all columns and rows

In [4]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# hide warnings
import warnings
warnings.filterwarnings('ignore')

- let's check the all detailed information of dataset through info function

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 231294 entries, 1999051 to 1763311
Data columns (total 24 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   App Name           231292 non-null  object 
 1   App Id             231294 non-null  object 
 2   Category           231294 non-null  object 
 3   Rating             229047 non-null  float64
 4   Rating Count       229047 non-null  float64
 5   Installs           231284 non-null  object 
 6   Minimum Installs   231284 non-null  float64
 7   Maximum Installs   231294 non-null  int64  
 8   Free               231294 non-null  bool   
 9   Price              231294 non-null  float64
 10  Currency           231281 non-null  object 
 11  Size               231280 non-null  object 
 12  Minimum Android    230662 non-null  object 
 13  Developer Id       231290 non-null  object 
 14  Developer Website  155451 non-null  object 
 15  Developer Email    231292 non-null  object 
 16  

# Observation:
1. total 24 columns of Data this dataset.
2. there are 4 boolean columns.
3. There is 1 integer column.
4. There are 4 float columns.
5. There are 15 object columns.

---
But we can completely rely on this formate of datast.

- check column names

In [12]:
df.columns

Index(['App Name', 'App Id', 'Category', 'Rating', 'Rating Count', 'Installs',
       'Minimum Installs', 'Maximum Installs', 'Free', 'Price', 'Currency',
       'Size', 'Minimum Android', 'Developer Id', 'Developer Website',
       'Developer Email', 'Released', 'Last Updated', 'Content Rating',
       'Privacy Policy', 'Ad Supported', 'In App Purchases', 'Editors Choice',
       'Scraped Time'],
      dtype='object')

- check top 10 values of the columns

In [20]:
df.head(10)

App Name  \
1999051                       Glückwünsche Zum Geburtstag   
1151748                                            Seu Zé   
538514                                      Keşifler Çağı   
522905                           radio lazer 102.9 oxnard   
1533904                                   Suarez Movement   
2180974                                      Trailer Cold   
1279455                               Supermercados Grilo   
1301021                            English Catechism Book   
795122   Christmas coloring book offline. Paint year 2021   
482870                                            Dot Box   

                                                App Id           Category  \
1999051              com.teamapps.happyBirthdaygermany          Lifestyle   
1151748                             br.com.bb.seuzeapp              Tools   
538514                             air.kesiflercagiand          Education   
522905   com.planetaapps.lazeroxnardonline.lazeroxnard      Music & Audio   
1533904                  com.trainerize.suarezmovement   Health & Fitness   
2180974                      com.flottaweb.trailercold       Productivity   
1279455                   com.VRSoft.SupermercadoGrilo           Shopping   
1301021                            com.moonstarinc.ecb  Books & Reference   
795122                   com.chrismasdraw.coloringbook      Entertainment   
482870                            com.BennyPink.DotBox             Casual   

         Rating  Rating Count Installs  Minimum Installs  Maximum Installs  \
1999051     0.0           0.0     100+             100.0               347   
1151748     NaN           NaN     500+             500.0               698   
538514      4.5           8.0     100+             100.0               435   
522905      0.0           0.0     100+             100.0               333   
1533904     0.0           0.0      10+              10.0                17   
2180974     0.0           0.0       1+               1.0                 2   
1279455     4.1          96.0  10,000+           10000.0             20831   
1301021     4.4         412.0  50,000+           50000.0             62406   
795122      5.0           5.0     100+             100.0               126   
482870      0.0           0.0       5+               5.0                 8   

         Free  Price Currency                Size     Minimum Android  \
1999051  True    0.0      USD                2.0M          4.2 and up   
1151748  True    0.0      USD  Varies with device  Varies with device   
538514   True    0.0      USD                5.1M          2.2 and up   
522905   True    0.0      USD                4.0M        4.0.3 and up   
1533904  True    0.0      USD                153M          6.0 and up   
2180974  True    0.0      USD                 27M          4.4 and up   
1279455  True    0.0      USD                 29M          4.1 and up   
1301021  True    0.0      USD                2.6M          4.1 and up   
795122   True    0.0      USD                 16M         4.4W and up   
482870   True    0.0      USD                 12M          4.1 and up   

                   Developer Id                     Developer Website  \
1999051                 CuteFun   http://www.meilleurjeuxandroid.net/   
1151748      Banco do Brasil SA                https://www.bb.com.br/   
538514   Terakki Vakfı Okulları             http://www.terakki.org.tr   
522905              planetaapps                                   NaN   
1533904              TRAINERIZE  http://trainwithluis.trainerize.com/   
2180974         Sistemisoft Srl            http://www.trailercold.com   
1279455              VR Connect                http://www.nweb.com.br   
1301021             MoonStarInc         https://www.moonstarinc.com/    
795122                 Tyler Ro                                   NaN   
482870        Byte Block Studio                                   NaN   

                       Developer Email      Released  Last Upd

# Observations:
- According to current observation, There are `10` columns which are numeric.
- These columns are `'Rating', 'Rating Count', 'Installs', 'Minimum Installs', 'Maximum Installs', 'Free', 'Size', 'Ad Supported', 'In App Purchases', 'Editors Choice'`

---
- Now, We know which are numeric columns and which columns need to be make consistant and correct the formating.

In [8]:
df.describe()

Rating  Rating Count  Minimum Installs  Maximum Installs  \
count  229047.000000  2.290470e+05      2.312840e+05      2.312940e+05   
mean        2.205521  2.558876e+03      2.094405e+05      3.700657e+05   
std         2.105733  1.028280e+05      1.886568e+07      3.398192e+07   
min         0.000000  0.000000e+00      0.000000e+00      0.000000e+00   
25%         0.000000  0.000000e+00      5.000000e+01      8.500000e+01   
50%         3.000000  6.000000e+00      5.000000e+02      6.930000e+02   
75%         4.300000  4.200000e+01      5.000000e+03      7.433000e+03   
max         5.000000  2.634006e+07      5.000000e+09      9.154248e+09   

               Price  
count  231294.000000  
mean        0.105997  
std         2.618078  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max       399.990000

## Observations:
- There are 4 columns are numeric.
- We need to observe more columns to verify.
- while the `'Rating', 'Rating Count', 'Minimum Installs', 'Maximum Installs', 'Price'` are the numeric.
- but, we don't know about the `Nan` values in these columns.

---
Let's check missing values of `'Rating'`, `'Rating Count'`, `'Minimum Installs'`, `'Maximum Installs'`, `'Price'`

In [27]:
df2[['Rating', 'Rating Count', 'Minimum Installs', 'Maximum Installs', 'Price']].isnull().sum().sort_values(ascending = False)
# print(f'Missing or Nan values of Rating are: ')

Rating              22883
Rating Count        22883
Minimum Installs      107
Maximum Installs        0
Price                   0
dtype: int64

# Observations:
- Missing values of `Rating` and`Rating Count` are `22883`.
- Missing values of `Minimum Installs` are `107`.
- while, missing value in `Maxim Installs` and `Price` are `0`.

- let's check column names

In [26]:
df.columns

Index(['App Name', 'App Id', 'Category', 'Rating', 'Rating Count', 'Installs',
       'Minimum Installs', 'Maximum Installs', 'Free', 'Price', 'Currency',
       'Size', 'Minimum Android', 'Developer Id', 'Developer Website',
       'Developer Email', 'Released', 'Last Updated', 'Content Rating',
       'Privacy Policy', 'Ad Supported', 'In App Purchases', 'Editors Choice',
       'Scraped Time'],
      dtype='object')

- let's check all columns missing values.

In [104]:
df2.isnull().sum().sort_values(ascending = False) 

Developer Website    760835
Privacy Policy       420953
Released              71053
Rating                22883
Rating Count          22883
Minimum Android        6530
Size                    196
Currency                135
Minimum Installs        107
Installs                107
Developer Id             33
Developer Email          31
App Name                  5
Maximum Installs          0
App Id                    0
Category                  0
Price                     0
Free                      0
Last Updated              0
Content Rating            0
Ad Supported              0
In App Purchases          0
Editors Choice            0
Scraped Time              0
dtype: int64

# Observations:
- Missing values of `'Developer Website'` are `760835`.
- Missing values of `'Privacy Policy'` are `420953`.
- Missing values of `'Released'` are `71053`.
- Missing values of `'Rating'` and `'Rating Count'` are `22883`.
- Missing values of `'Minimum Android'` are `6530`.
- Missing values of `'Size'` are `196`.
- Missing values of `'Currency'` are `135`.
- Missing values of `'Minimum Installs'`, `'Installs'` are `107`.
- Missing values of `'Developer Id'` are `33`.
- Missing values of `'Developer Email'` are `31`.
- Missing values of `'App Name'` are `5`.
- Missing values of `'Maximum Installs'`, `'App Id'`, `'Category'`, `'Price'`, `'Free'`, `'Last Updated'`, `'Content Rating'`, `'Ad Supported'`, `'In App Purchases'`, `'Editors choice'`, `'Scraped Time'` are `0`.


- let's check the percentage of missing values.

In [78]:
df.columns[1]

'App Id'

In [103]:
cols = df.columns

# def null_values(col):

for col in cols:
#     # print(col)
#     print()
    if df2[col].isnull().sum() > 0:
        print(f'Missing values in percentage of {col} :  {round(df2[col].isnull().sum()/len(df2) * 100 , 5)}')
    
# col

Missing values in percentage of App Name :  0.00022
Missing values in percentage of Rating :  0.98935
Missing values in percentage of Rating Count :  0.98935
Missing values in percentage of Installs :  0.00463
Missing values in percentage of Minimum Installs :  0.00463
Missing values in percentage of Currency :  0.00584
Missing values in percentage of Size :  0.00847
Missing values in percentage of Minimum Android :  0.28232
Missing values in percentage of Developer Id :  0.00143
Missing values in percentage of Developer Website :  32.89466
Missing values in percentage of Developer Email :  0.00134
Missing values in percentage of Released :  3.07197
Missing values in percentage of Privacy Policy :  18.19988


# Observations:
Missing values in percentage of App Name :  0.00022\
Missing values in percentage of Rating :  0.98935\
Missing values in percentage of Rating Count :  0.98935\
Missing values in percentage of Installs :  0.00463\
Missing values in percentage of Minimum Installs :  0.00463\
Missing values in percentage of Currency :  0.00584\
Missing values in percentage of Size :  0.00847\
Missing values in percentage of Minimum Android :  0.28232\
Missing values in percentage of Developer Id :  0.00143\
Missing values in percentage of Developer Website :  32.89466\
Missing values in percentage of Developer Email :  0.00134\
Missing values in percentage of Released :  3.07197\
Missing values in percentage of Privacy Policy :  18.19988


---
let's 